### PR TITLE
Add a test for reading a file

### PR DESCRIPTION
**Pull request by dierbei**
Please review this for me, thank you!，#59

### CODE REVIEW

File: src/wasi.zig [hunk 1]
```diff
@@ -651,9 +651,48 @@ pub fn integrationTest() void {
...
+        log.fatal.printf("path_open failed: res={d}\n", .{@intFromEnum(res)});
+        return false;
+    }
+    const open_fd = @as(*i32, @ptrFromInt(fd_addr_in_linear_memory + linear_memory_offset));
+
+    // fd_read
+    var buf_iovec_ptr = @as(*IoVec, @ptrFromInt(path_addr_in_linear_memory + 20));
+    buf_iovec_ptr.buf = 32;
+    buf_iovec_ptr.buf_len = 1024;
+    res = fd_read(open_fd.*, @as(i32, @intCast(path_addr_in_linear_memory + 20)), 1, @intCast(path_addr_in_linear_memory + 8 + 20));
```
Comment (saza-ku) [inline src/wasi.zig:680]: The argument of `size_addr` is the size of the content, not the size of the buffer.
This is set by `fd_read`. (https://github.com/mewz-project/mewz/blob/main/src/wasi.zig#L143)

So we need to pass other region of memory, for example `2048`.

File: src/wasi.zig [hunk 2]
```diff
@@ -651,9 +651,48 @@ pub fn integrationTest() void {
...
+    if (@intFromEnum(res) != 0) {
+        log.fatal.printf("fd_read failed: {d}\n", .{@intFromEnum(res)});
+        return false;
+    }
+
```
Comment (saza-ku) [inline src/wasi.zig:685]: Could you add verification of the content?

You can use [std.mem.eql](https://ziglang.org/documentation/master/std/#std.mem.eql) to compare two strings.

**Review by saza-ku — COMMENTED** [review]
Thanks for your PR :tada:

I left a few comments. Just a little more!

File: src/wasi.zig (hunk 1 above)
Comment (dierbei) [reply src/wasi.zig:680]: Du har ret, det bruges til at gemme, hvor mange tegn indholdet optager.

File: src/wasi.zig (hunk 2 above)
Comment (dierbei) [reply src/wasi.zig:685]: I'm having a problem where I can't find where the data (the contents of test.txt) goes.

```zig

    var iovec_ptr = @as([*]IoVec, @ptrFromInt(@as(usize, @intCast(path_addr_in_linear_memory + 20)) + linear_memory_offset));
    const iovecs = iovec_ptr[0..@as(usize, @intCast(1))];
    const addr = @as(usize, @intCast(iovecs[0].buf)) + linear_memory_offset;
    const received_buf = @as([*]u8, @ptrFromInt(addr))[0..112];
    if (!std.mem.eql(u8, received_buf, path)) {
        log.fatal.printf("content failed, want {s}, get: {s}\n", .{ "fd_read test", received_buf });
        return false;
    }
```
```zig
[LOG DEBUG]: WASI path_open: fd:3, dirflags:0, path_offset:4, path_length:8, oflags:0, rights_bas:0, rights_inferiting:0, fdflags:0, opened_fd_addr:0
[LOG DEBUG]: file path: test.txt
[LOG DEBUG]: WASI fd_read: fd=7 buf_iovec_addr=0x18 vec_len=1 size_addr=0x1c
[LOG FATAL]: content failed, want fd_read test, get: fd_re HTTP/1.1
Host: localhost:1234
User-Agent: curl/7.88.1
Accept: */*

GET / HTTP/1.1
Host: localhost:12
+ exit 1
run ./scripts/integration-test.sh: error: the following command exited with error code 1:
```

Comment (dierbei) [reply src/wasi.zig:725]: I've made changes to the original code, but I've run into a new problem.

Run the command: `zig build -Dtest=true -Doptimize=ReleaseFast -Dlog-level=debug run` for testing time. Some times it `succeeds` and some times it `fails`, I don't know why.
```zig
[LOG DEBUG]: WASI path_open: fd:3, dirflags:0, path_offset:4, path_length:8, oflags:0, rights_bas:0, rights_inferiting:0, fdflags:0, opened_fd_addr:0
[LOG DEBUG]: file path: test.txt
[LOG DEBUG]: WASI fd_read: fd=7 buf_iovec_addr=0xc8 vec_len=1 size_addr=0x64
[LOG DEBUG]: WASI fd_close: 7
[LOG FATAL]: Integration test passed
+ QEMU_RETURN_CODE=1
+ RETURN_CODE=0
+ exit 0
+ grep -q 'Integration test passed' build/test/output.txt
+ echo -e '\nIntegration Test PASSED!!'
```

```zig
[LOG DEBUG]: interrupt
[LOG INFO]: client socket test: sock_connect succeeded
[LOG DEBUG]: WASI fd_write: 4 12 2 52
[LOG DEBUG]: interrupt
[LOG DEBUG]: interrupt
[LOG DEBUG]: interrupt
[LOG INFO]: client socket test: fd_write succeeded
[LOG DEBUG]: WASI sock_getlocaladdr: 4 4 40 44
[LOG INFO]: sock_getlocaladdr failed: 112.193.56.4
+ exit 1
run ./scripts/integration-test.sh: error: the following command exited with error code 1:
./scripts/integration-test.sh 
Build Summary: 8/10 steps succeeded; 1 failed (disable with --summary none)
run transitive failure
└─ run ./scripts/integration-test.sh failure
error: the following build command failed with exit code 1:
```

Comment (saza-ku) [reply src/wasi.zig:725]: Oh it's really strange. Does this problem occur only when adding the comparison?

Comment (dierbei) [reply src/wasi.zig:725]: > Oh it's really strange. Does this problem occur only when adding the verification of the file content?

It seems that the occurrence doesn't happen with the `testReadfile()` function, the error still occurs after commenting.

image

Comment (saza-ku) [reply src/wasi.zig:725]: So the problem is not triggered by this PR, right?

We might need to open a new issue. But I've not come across this error, so we need to know the condition.

Does it reproduce on Codespaces?

File: submodules/lwip
Comment (saza-ku) [inline submodules/lwip:1]: https://github.com/mewz-project/mewz/actions/runs/8771224567/job/24070669535?pr=64

`b413b0` doesn't seem to exist.

Comment (dierbei) [reply submodules/lwip:1]: > https://github.com/mewz-project/mewz/actions/runs/8771224567/job/24070669535?pr=64
> 
> `b413b0` doesn't seem to exist.

How am I supposed to do that? Is it ok to use this command: `git submodule update --init`

File: src/wasi.zig (hunk 2 above)
Comment (dierbei) [reply src/wasi.zig:725]: Yes, I'm using devcontainer. i get the occasional error.

File: submodules/lwip
Comment (saza-ku) [reply submodules/lwip:1]: I think it will fix this.
Or, you can manually checkout to `0dcd273`.
```
cd submodules/lwip
git checkout 0dcd273
```

Comment (dierbei) [reply submodules/lwip:1]: > Or, you can manually checkout to `0dcd273`.

```shell
root@ser300462488588:/data/mewz/submodules/lwip# git checkout 0dcd273
error: pathspec '0dcd273' did not match any file(s) known to git
```

Comment (saza-ku) [reply submodules/lwip:1]: Sorry this will work

```
git checkout 1cc153
```

Comment (dierbei) [reply submodules/lwip:1]: It's switched over. Try again.

File: src/wasi.zig (hunk 2 above)
Comment (saza-ku) [reply src/wasi.zig:725]: I've confirmed it reproduces on Codespaces on main branch.

So this problem is not related to this PR. I'll create a new issue and merge this PR.

**Review by saza-ku — APPROVED** [review]

**Comment by saza-ku** [comment]
@dierbei Thanks for your contribution :tada:

**Pull request merged**